### PR TITLE
Basic Structure of bigdl-nano

### PR DESCRIPTION
**Pull request by leonardozcm**
```
- python/
   - nano/
      - src/
      |  - bigdl/
      |    - nano/ 
      |      - pytorch/
      |         - trainer/
      |         - vision/
      |      - tf/
      |         - keras/
      |      - common/
      - script/
      - recipe/
      - setup.py
```

### CODE REVIEW

**Comment by leonardozcm** [comment]
setup.py will pack bigdl module, and module named `orca_lite` will be replaced to `bigdl.nano`

File: python/nano/recipe/BuildGuidance.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+conda install conda-build
+```
+## Conda Recipe
+Details at https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html
+
+## Build
+For example, we plan to build a package based on python3.7
+1. Just run:
+    ```bash
+    cd orca-lite-poc
```
Comment (jason-dai) [inline python/nano/recipe/BuildGuidance.md:16]: bigdl-nano

File: python/nano/recipe/activate.sh [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash -e -x
+
+echo "=====activate===="
+is_tf = $(conda list orca-lite.tf | grep orca-lite.tf) || is_tf = "no"
```
Comment (jason-dai) [inline python/nano/recipe/activate.sh:4]: bigdl-nano

File: python/nano/recipe/activate.sh [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+if [ "${is_tf}" = "no" ];then
+    echo "Use orca-lite-init support for pytorch"
```
Comment (jason-dai) [inline python/nano/recipe/activate.sh:7]: bigdl-nano

File: python/nano/recipe/activate.sh [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+    nano-init
+else
+    echo "find orca-lite-init support for tensorflow"
```
Comment (jason-dai) [inline python/nano/recipe/activate.sh:10]: bigdl-nano

File: python/nano/recipe/BuildGuidance.md (hunk 1 above)
Comment (leonardozcm) [reply python/nano/recipe/BuildGuidance.md:16]: ok, our recipe file will be update after https://github.com/analytics-zoo/orca-lite-poc/pull/159

**Comment by leonardozcm** [comment]
Conda packages are available now, try `conda install bigdl-nano -c chriskafka -c conda-forge -c pytorch`
while bigdl-nano contains empty context.

**Comment by leonardozcm** [comment]
```
(bigdlnanotest38) root@3659e767d1ab:~/analytics-zoo/python/nano# conda install bigdl-nano -c chriskafka -c conda-forge -c pytorch
Collecting package metadata (current_repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/bigdlnanotest38

  added / updated specs:
    - bigdl-nano


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    bigdl-nano-0.0.1           |             py_0           4 KB  chriskafka
    bigdl-nano.base-0.0.1      |             py_0          11 KB  chriskafka
    bigdl-nano.pytorch-0.0.1   |             py_0           4 KB  chriskafka
    ------------------------------------------------------------
                                           Total:          19 KB

The following NEW packages will be INSTALLED:

  absl-py            conda-forge/noarch::absl-py-0.13.0-pyhd8ed1ab_0
  aiohttp            conda-forge/linux-64::aiohttp-3.7.4.post0-py38h497a2fe_0
  async-timeout      conda-forge/noarch::async-timeout-3.0.1-py_1000
  attrs              conda-forge/noarch::attrs-21.2.0-pyhd8ed1ab_0
  bigdl-nano         chriskafka/noarch::bigdl-nano-0.0.1-py_0
  bigdl-nano.base    chriskafka/noarch::bigdl-nano.base-0.0.1-py_0
  bigdl-nano.pytorch chriskafka/noarch::bigdl-nano.pytorch-0.0.1-py_0
  blas               pkgs/main/linux-64::blas-1.0-mkl
  blinker            conda-forge/noarch::blinker-1.4-py_1
  brotlipy           conda-forge/linux-64::brotlipy-0.7.0-py38h497a2fe_1001
  bzip2              conda-forge/linux-64::bzip2-1.0.8-h7f98852_4
  c-ares             pkgs/main/linux-64::c-ares-1.17.1-h27cfd23_0
  cachetools         conda-forge/noarch::cachetools-4.2.2-pyhd8ed1ab_0
  cffi               conda-forge/linux-64::cffi-1.14.6-py38ha65f79e_0
  chardet            conda-forge/linux-64::chardet-4.0.0-py38h578d9bd_1
  charset-normalizer conda-forge/noarch::charset-normalizer-2.0.0-pyhd8ed1ab_0
  click              conda-forge/linux-64::click-8.0.1-py38h578d9bd_0
  colorama           conda-forge/noarch::colorama-0.4.4-pyh9f0ad1d_0
  cpuonly            pytorch/noarch::cpuonly-1.0-0
  cryptography       conda-forge/linux-64::cryptography-3.4.7-py38ha5dfef3_0
  dataclasses        conda-forge/noarch::dataclasses-0.8-pyhc8e2a94_1
  ffmpeg             conda-forge/linux-64::ffmpeg-4.3.2-hca11adc_0
  freetype           conda-forge/linux-64::freetype-2.10.4-h0708190_1
  fsspec             conda-forge/noarch::fsspec-2021.7.0-pyhd8ed1ab_0
  future             conda-forge/linux-64::future-0.18.2-py38h578d9bd_3
  gmp                conda-forge/linux-64::gmp-6.2.1-h58526e2_0
  gnutls             conda-forge/linux-64::gnutls-3.6.13-h85f3911_1
  google-auth        conda-forge/noarch::google-auth-1.34.0-pyh6c4a22f_0
  google-auth-oauth~ conda-forge/noarch::google-auth-oauthlib-0.4.1-py_2
  grpcio             conda-forge/linux-64::grpcio-1.38.1-py38hdd6454d_0
  idna               conda-forge/noarch::idna-3.1-pyhd3deb0d_0
  importlib-metadata conda-forge/linux-64::importlib-metadata-4.6.3-py38h578d9bd_0
  intel-openmp       pkgs/main/linux-64::intel-openmp-2021.3.0-h06a4308_3350
  intel_pytorch_ext~ chriskafka/linux-64::intel_pytorch_extension-1.8.0-py38_0
  jemalloc           chriskafka/linux-64::jemalloc-5.2.1-h9c3ff4c_5
  jpeg               pkgs/main/linux-64::jpeg-9b-h024ee3a_2
  lame               conda-forge/linux-64::lame-3.100-h7f98852_1001
  lcms2              pkgs/main/linux-64::lcms2-2.12-h3be6417_0
  libjpeg-turbo      conda-forge/linux-64::libjpeg-turbo-2.1.0-h7f98852_0
  libpng             conda-forge/linux-64::libpng-1.6.37-h21135ba_2
  libprotobuf        conda-forge/linux-64::libprotobuf-3.17.2-h780b84a_1
  libtiff            pkgs/main/linux-64::libtiff-4.2.0-h85742a9_0
  libuv              conda-forge/linux-64::libuv-1.42.0-h7f98852_0
  libwebp-base       conda-forge/linux-64::libwebp-base-1.2.0-h7f98852_2
  lz4-c              conda-forge/linux-64::lz4-c-1.9.3-h9c3ff4c_1
  markdown           conda-forge/noarch::markdown-3.3.4-pyhd8ed1ab_0
  mkl                pkgs/main/linux-64::mkl-2021.3.0-h06a4308_520
  mkl-service        conda-forge/linux-64::mkl-service-2.4.0-py38h497a2fe_0
  mkl_fft            pkgs/main/linux-64::mkl_fft-1.3.0-py38h42c9631_2
  mkl_random         conda-forge/linux-64::mkl_random-1.2.2-py38h1abd341_0
  multidict          conda-forge/linux-64::multidict-5.1.0-py38h497a2fe_1
  nettle             conda-forge/linux-64::nettle-3.6-he412f7d_0
  ninja              conda-forge/linux-64::ninja-1.10.2-h4bd325d_0
  numpy              pkgs/main/linux-64::numpy-1.20.3-py38hf144106_0
  numpy-base         pkgs/main/linux-64::numpy-base-1.20.3-py38h74d4b33_0
  oauthlib           conda-forge/noarch::oauthlib-3.1.1-pyhd8ed1ab_0
  olefile            conda-forge/noarch::olefile-0.46-pyh9f0ad1d_1
  opencv-python-hea~ chriskafka/linux-64::opencv-python-headless-4.5.3.56-py38_0
  opencv_transforms  chriskafka/noarch::opencv_transforms-0.0.1-py_0
  openh264           conda-forge/linux-64::openh264-2.1.1-h780b84a_0
  openjpeg           conda-forge/linux-64::openjpeg-2.4.0-hb52868f_1
  packaging          conda-forge/noarch::packaging-21.0-pyhd8ed1ab_0
  pillow             pkgs/main/linux-64::pillow-8.3.1-py38h2c7a002_0
  protobuf           conda-forge/linux-64::protobuf-3.17.2-py38h709712a_0
  pyasn1             conda-forge/noarch::pyasn1-0.4.8-py_0
  pyasn1-modules     conda-forge/noarch::pyasn1-modules-0.2.7-py_0
  pycparser          conda-forge/noarch::pycparser-2.20-pyh9f0ad1d_2
  pydeprecate        conda-forge/noarch::pydeprecate-0.3.1-pyhd8ed1ab_0
  pyjwt              conda-forge/noarch::pyjwt-2.1.0-pyhd8ed1ab_0
  pyopenssl          conda-forge/noarch::pyopenssl-20.0.1-pyhd8ed1ab_0
  pyparsing          conda-forge/noarch::pyparsing-2.4.7-pyh9f0ad1d_0
  pysocks            conda-forge/linux-64::pysocks-1.7.1-py38h578d9bd_3
  python_abi         conda-forge/linux-64::python_abi-3.8-2_cp38
  pytorch            pytorch/linux-64::pytorch-1.8.0-py3.8_cpu_0
  pytorch-lightning  conda-forge/noarch::pytorch-lightning-1.4.2-pyhd8ed1ab_0
  pyturbojpeg        conda-forge/noarch::pyturbojpeg-1.5.1-pyhec5db9e_0
  pyu2f              conda-forge/noarch::pyu2f-0.1.5-pyhd8ed1ab_0
  pyyaml             conda-forge/linux-64::pyyaml-5.4.1-py38h497a2fe_0
  requests           conda-forge/noarch::requests-2.26.0-pyhd8ed1ab_0
  requests-oauthlib  conda-forge/noarch::requests-oauthlib-1.3.0-pyh9f0ad1d_0
  rsa                conda-forge/noarch::rsa-4.7.2-pyh44b312d_0
  six                conda-forge/noarch::six-1.16.0-pyh6c4a22f_0
  tensorboard        conda-forge/noarch::tensorboard-2.6.0-pyhd8ed1ab_0
  tensorboard-data-~ conda-forge/linux-64::tensorboard-data-server-0.6.0-py38h2b97feb_0
  tensorboard-plugi~ conda-forge/noarch::tensorboard-plugin-wit-1.8.0-pyh44b312d_0
  torchmetrics       conda-forge/noarch::torchmetrics-0.5.0-pyhd8ed1ab_0
  torchvision        pytorch/linux-64::torchvision-0.9.0-py38_cpu
  tqdm               conda-forge/noarch::tqdm-4.62.0-pyhd8ed1ab_0
  typing-extensions  conda-forge/noarch::typing-extensions-3.10.0.0-hd8ed1ab_0
  typing_extensions  conda-forge/noarch::typing_extensions-3.10.0.0-pyha770c72_0
  urllib3            conda-forge/noarch::urllib3-1.26.6-pyhd8ed1ab_0
  werkzeug           conda-forge/noarch::werkzeug-2.0.1-pyhd8ed1ab_0
  x264               conda-forge/linux-64::x264-1!161.3030-h7f98852_1
  yaml               conda-forge/linux-64::yaml-0.2.5-h516909a_0
  yarl               conda-forge/linux-64::yarl-1.6.3-py38h497a2fe_2
  zipp               conda-forge/noarch::zipp-3.5.0-pyhd8ed1ab_0
  zstd               conda-forge/linux-64::zstd-1.4.9-ha95c52a_0

The following packages will be SUPERSEDED by a higher-priority channel:

  ca-certificates    pkgs/main::ca-certificates-2021.7.5-h~ --> conda-forge::ca-certificates-2021.5.30-ha878542_0
  certifi            pkgs/main::certifi-2021.5.30-py38h06a~ --> conda-forge::certifi-2021.5.30-py38h578d9bd_0
  openssl              pkgs/main::openssl-1.1.1k-h27cfd23_0 --> conda-forge::openssl-1.1.1k-h7f98852_0


Proceed ([y]/n)? 


Downloading and Extracting Packages
bigdl-nano-0.0.1     | 4 KB      | ################################################################################################################################################################################################## | 100% 
bigdl-nano.pytorch-0 | 4 KB      | ################################################################################################################################################################################################## | 100% 
bigdl-nano.base-0.0. | 11 KB     | ################################################################################################################################################################################################## | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
Option: Disable opemMP and unset related variables
Option: Disable jemalloc and unset related variables
conda dir found: /opt/conda/envs/bigdlnanotest38
OpenMP library found...
Setting OMP_NUM_THREADS...
Setting OMP_NUM_THREADS specified for pytorch...
Setting KMP_AFFINITY...
Setting KMP_BLOCKTIME...
Jemalloc library found...
Setting MALLOC_CONF...
Setting LD_PRELOAD...
+++++ Env Variables +++++
LD_PRELOAD=
MALLOC_CONF=
OMP_NUM_THREADS=
KMP_AFFINITY=
KMP_BLOCKTIME=
TF_ENABLE_ONEDNN_OPTS=
+++++++++++++++++++++++++
Complete.
=====activate====
Use bigdl-nano-run support for pytorch
conda dir found: /opt/conda/envs/bigdlnanotest38
OpenMP library found...
Setting OMP_NUM_THREADS...
Setting OMP_NUM_THREADS specified for pytorch...
Setting KMP_AFFINITY...
Setting KMP_BLOCKTIME...
Jemalloc library found...
Setting MALLOC_CONF...
Setting LD_PRELOAD...
+++++ Env Variables +++++
LD_PRELOAD=/opt/conda/envs/bigdlnanotest38/lib/libiomp5.so /opt/conda/envs/bigdlnanotest38/lib/libjemalloc.so
MALLOC_CONF=oversize_threshold:1,background_thread:true,metadata_thp:auto,dirty_decay_ms:-1,muzzy_decay_ms:-1
OMP_NUM_THREADS=48
KMP_AFFINITY=granularity=fine,compact,1,0
KMP_BLOCKTIME=1
TF_ENABLE_ONEDNN_OPTS=
+++++++++++++++++++++++++
Complete.
(bigdlnanotest38) root@3659e767d1ab:~/analytics-zoo/python/nano# python -c "import bigdl.nano"
```

**Comment by jason-dai** [comment]
Why do we have three packages?

```
    package                    |            build
    ---------------------------|-----------------
    bigdl-nano-0.0.1           |             py_0           4 KB  chriskafka
    bigdl-nano.base-0.0.1      |             py_0          11 KB  chriskafka
    bigdl-nano.pytorch-0.0.1   |             py_0           4 KB  chriskafka
    ------------------------------------------------------------
```

**Comment by leonardozcm** [comment]
> Why do we have three packages?
> 
> ```
>     package                    |            build
>     ---------------------------|-----------------
>     bigdl-nano-0.0.1           |             py_0           4 KB  chriskafka
>     bigdl-nano.base-0.0.1      |             py_0          11 KB  chriskafka
>     bigdl-nano.pytorch-0.0.1   |             py_0           4 KB  chriskafka
>     ------------------------------------------------------------
> ```

As scheduled in meta.yaml:
https://github.com/intel-analytics/analytics-zoo/blob/71ed23d9732013ed95d10e482da9eb62f07f9aab/python/nano/recipe/meta.yaml#L104
bigdl-nano-0.0.1 only contains bigdl-nano.pytorch-0.0.1 as its dependency,  bigdl-nano.base serves as some base package used by  bigdl-nano.pytorch and  bigdl-nano.tf. 
Could we upload all these package to Anaconda.org while just display bigdl-nano,  bigdl-nano.pytorch and  bigdl-nano.tf as available?

File: python/nano/setup.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        print(dirpath)
+        package = dirpath.split(nano_home + "/")[1].replace('/', '.')
+        if any(fnmatch.fnmatchcase(package, pat=pattern)
+                for pattern in exclude_patterns):
+            print("excluding", package)
+        else:
+            nano_packages.append(package)
+            print("including", package)
+    return nano_packages
+print(get_nano_packages())
```
Comment (yangw1234) [inline python/nano/setup.py:24]: remove this line?

File: python/nano/script/bigdl-nano-run [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+echo "LD_PRELOAD=${LD_PRELOAD}"
+echo "MALLOC_CONF=${MALLOC_CONF}"
+echo "OMP_NUM_THREADS=${OMP_NUM_THREADS}"
+echo "KMP_AFFINITY=${KMP_AFFINITY}"
+echo "KMP_BLOCKTIME=${KMP_BLOCKTIME}"
+echo "TF_ENABLE_ONEDNN_OPTS=${TF_ENABLE_ONEDNN_OPTS}"
+echo "+++++++++++++++++++++++++"
+# Run the commands after bigdl-nano-run
+echo "Complete."
+${@:1}
```
Comment (yangw1234) [inline python/nano/script/bigdl-nano-run:195]: we do not need this line anymore right?

**Comment by yangw1234** [comment]
@jason-dai should we first merge this, so that further development can move to this branch?

**Comment by jason-dai** [comment]
> @jason-dai should we first merge this, so that further development can move to this branch?

LGTM

File: python/nano/setup.py (hunk 5 above)
Comment (leonardozcm) [reply python/nano/setup.py]: ok, fixed

File: python/nano/script/bigdl-nano-run (hunk 6 above)
Comment (leonardozcm) [reply python/nano/script/bigdl-nano-run]: fixed

**Comment by yangw1234** [comment]
@jason-dai about using `import bigdl.nano` or `import bigdl.orca`, I have some concerns.

We have to include a `bigdl` python module in both "bigdl-nano" and "bigdl-orca" package if supporting this. If both the project define a `bigdl/__init__.py`, then last installed package will override the previous one.

Do you think it is going to be a problem?

Should we use `import bigdl-nano`? 
Like pytorch, they are using `import torchvision` but not `import pytorch.vision`. And in tensorflow, using `import tensorflow_datasets` not `tensorflow.datasets`

**Comment by jason-dai** [comment]
> @jason-dai about using `import bigdl.nano` or `import bigdl.orca`, I have some concerns.
> 
> We have to include a `bigdl` python module in both "bigdl-nano" and "bigdl-orca" package if supporting this. If both the project define a `bigdl/__init__.py`, then last installed package will override the previous one.
> 
> Do you think it is going to be a problem?
> 
> Should we use `import bigdl-nano`?
> Like pytorch, they are using `import torchvision` but not `import pytorch.vision`. And in tensorflow, using `import tensorflow_datasets` not `tensorflow.datasets`

If we make sure `bigdl/__init__.py` is empty, is this still a problem?

**Comment by yangw1234** [comment]
> > @jason-dai about using `import bigdl.nano` or `import bigdl.orca`, I have some concerns.
> > We have to include a `bigdl` python module in both "bigdl-nano" and "bigdl-orca" package if supporting this. If both the project define a `bigdl/__init__.py`, then last installed package will override the previous one.
> > Do you think it is going to be a problem?
> > Should we use `import bigdl-nano`?
> > Like pytorch, they are using `import torchvision` but not `import pytorch.vision`. And in tensorflow, using `import tensorflow_datasets` not `tensorflow.datasets`
> 
> If we make sure `bigdl/__init__.py` is empty, is this still a problem? And I would assume that the project structure will be
> 
> ```
> python\bigdl\dllib
> python\bigdl\nano
> python\bigdl\orca
> ```
> 
> And we should have only one `bigdl/__init__.py`
Yes, I think making sure `bigdl/__init__.py` is empty and having only one `bigdl/__init__.py` would solve the problem. 

If we structure the project this way, there maybe other things to consider 
e.g. 
1. code in bigdl submodules does not call each other. e.g. code in `bigdl.orca` cannot call a function defined in `bigdl.nano`. In the development environment, this works. It does not work if user only install one package.
2. no code in `bigdl/` level, otherwise, installing difference versions submoduels might cause problems. e.g. install bigdl-orca==2.1.0 and bigdl-orca==2.2.0
3. what if user first bigdl-dllib==2.0.0, then install bigdl==0.13.0. The old bigdl will put a lot things in the `bigdl/` directory, including `__init__.py`, not sure it will work.

**Pull request merged**